### PR TITLE
[FSSDK-9471] [React] Fetch segments in onready

### PR DESCRIPTION
**Pull request by rafinutshaw-optimizely**
## Summary
- Add fetchqualifiedsegments options params to onReady client method
- Fetch qualified segments in onready if usercontext is available
- 
## Test plan
- fetchqualifiedsegment method is called in onReady

## Issues
- https://jira.sso.episerver.net/browse/FSSDK-9471

### CODE REVIEW

File: src/client.spec.ts [hunk 1]
```diff
@@ -142,6 +142,15 @@ describe('ReactSDKClient', () => {
       expect(result.success).toBe(false);
     });
 
+    it('calls fetchqualifiedsegments when a user is set', async () => {
```
Comment (mikechu-optimizely) [inline src/client.spec.ts:145]: Should we add some additional tests for the negative cases like 
- when we don't have a userContext or 
- when there's a timeout and we verify that there was a `message` for 'failed to initialize onReady before timeout, either the datafile, user info or qualified segments was not set before the timeout'

File: src/client.ts [hunk 2]
```diff
@@ -281,26 +281,28 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
...
-            'failed to initialize onReady before timeout, either the datafile or user info was not set before the timeout',
+            'failed to initialize onReady before timeout, either the datafile, user info or qualified segments was not set before the timeout',
           dataReadyPromise: this.dataReadyPromise,
         });
       }, timeout) as any;
     });
 
-    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(res => {
+    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(async res => {
+      await this.fetchQualifiedSegments(config?.segmentOptions);
```
Comment (mikechu-optimizely) [inline src/client.ts:305]: Do we need testing around a failure for `fetchQualifiedSegments()`?

Comment (mikechu-optimizely) [reply src/client.ts:305]: ...or maybe we already cover it.

**Review by mikechu-optimizely — APPROVED** [review]
LGTM, with no showstoppers from my perspective.  I have a couple of questions below.

File: src/client.ts (hunk 2 above)
Comment (jaeopt) [inline src/client.ts:305]: `fetchQualifiedSegment` should be called only on datafile success not on timeout. Also can you check if this the caller is blocked until `fetchQualifiedSegments` completed. It looks like async call.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Questions about fetch timing.

File: src/client.spec.ts (hunk 1 above)
Comment (rafinutshaw-optimizely) [reply src/client.spec.ts:145]: - we already have a userContext check inside fetchqualified method
- Updated changes for this

File: src/client.ts (hunk 2 above)
Comment (rafinutshaw-optimizely) [reply src/client.ts]: yes failure for fetchQualifiedSegments has been cover on fetchQualifiedSegments method itself. So we can just fire and forget

Comment (rafinutshaw-optimizely) [reply src/client.ts]: Thanks. added a check timeout check on this. And yes this the caller is blocked until fetchQualifiedSegments completed

File: src/client.ts [hunk 3]
```diff
@@ -294,13 +296,14 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
...
-            'failed to initialize onReady before timeout, either the datafile or user info was not set before the timeout',
+            'failed to initialize onReady before timeout, either the datafile, user info or qualified segments was not set before the timeout',
           dataReadyPromise: this.dataReadyPromise,
         });
       }, timeout) as any;
     });
 
-    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(res => {
+    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(async res => {
+      if (res.success) await this.fetchQualifiedSegments(config?.segmentOptions);
```
Comment (mikechu-optimizely) [inline src/client.ts:306]: Can we use two lines for this and enclose in { } ?

**Review by mikechu-optimizely — APPROVED** [review]
LGTM

File: src/client.ts [hunk 4]
```diff
@@ -294,13 +296,16 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
           success: false,
           reason: 'TIMEOUT',
           message:
-            'failed to initialize onReady before timeout, either the datafile or user info was not set before the timeout',
+            'failed to initialize onReady before timeout, either the datafile, user info or qualified segments was not set before the timeout',
```
Comment (jaeopt) [inline src/client.ts:299]: I believe this timeout is covering datafile ready timeout, not covering "qualifed segments" failure. The original message makes more sense.

File: src/client.ts [hunk 5]
```diff
@@ -294,13 +296,16 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
...
           dataReadyPromise: this.dataReadyPromise,
         });
       }, timeout) as any;
     });
 
-    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(res => {
+    return Promise.race([this.dataReadyPromise, timeoutPromise]).then(async res => {
+      if (res.success) {
+        await this.fetchQualifiedSegments(config?.segmentOptions);
```
Comment (jaeopt) [inline src/client.ts:307]: What if fetch fails with network error, timeout, etc? res should be updated accordingly?

**Review by jaeopt — CHANGES_REQUESTED** [review]
The flag looks good. A couple of more questions.

**Comment by rafinutshaw-optimizely** [comment]
Closing this due to change in requirement.
